### PR TITLE
Marketplace: Add "Add to Store" button for free and WordPress.org products

### PR DESCRIPTION
**Pull request by raicem**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

- Allows merchants to install any free and .org hosted product from any page, while creating the order on the go. Currently this mocks order creation process.
- Adds new `<InstallNewProductModal>` component. This listens for changes in `query`, and shows the corresponding modal to the user.
- Adds new **Add to Store** button on the search pages. You need to switch to the Woo.com branch as well.

### How to test the changes in this Pull Request:

1. Connect front end and back end to your local Woo.com instance. 
    - For front end, switch `MARKETPLACE_HOST` in `marketplace/components/constants.ts` to `https://woocommerce.test`.
    - For back-end, make modifications in `class-wc-helper`. Make sure you also connect to the right Docker network.
2. Build assets with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build`
3. Disconnect store from Woo.com by clicking **Avatar > Disconnect** from store
    - Please note which page you initiate the disconnection from. After the disconnect is done, you should be redirected back to the same page.
5. [On the search page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=extensions&path=%2Fextensions), find a product with the **Add to Store** button and click it. It should show a warning for requiring a connection.
6. Complete the connection. Please note when you come back to your store, the modal should open automatically.
7. In the next modal, click Install. This will first ping Woo.com, create an order and then start downloading the product from WordPress.org repository. 
8. When the installation is done and we're allowed to activate (sometimes we're not allowed), you should see the **Activate** button. Clicking the button will activate the product and then show an option to see the plugin list, to set up the plugin. 
9. On Woo.com, delete your customer's billing address and try to install the plugin again. You won't see the **Add to Store** on the same product again, because it's now installed. 
10. Trying to install a new product, should first show a warning and then redirect user to Woo.com with the product added to cart.
11. When it's done, we should test installations with a product key. Find a product key from the API requests in [the My Subscriptions page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions) and then visit http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions&install=PRODUCT_KEY_STARTS_WITH_W00.
12. On the [the My Subscriptions page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions) click **Install** in the table and confirm it works.

### CODE REVIEW

File: plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+					throw response;
+				}
+
+				const subscriptionData = response.data.subscription;
+
+				setSubscription( subscriptionData );
+
+				dispatch( installingStore ).startInstalling(
+					subscriptionData.product_key
+				);
```
Comment (raicem) [inline plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx]: Do you think I need to dispatch this to the store @bgrgicak?

File: plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx
Comment (raicem) [inline plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx:1]: This file handles the whole flow from start to finish. Therefore, in the end, I didn't need to change anything from the previous installation flows. 

I think file is a little too big. Maybe I'll refactor it down the line, when making tweaks on the feature branch.

File: plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+import { dispatch } from '@wordpress/data';
+import { useState, useEffect } from '@wordpress/element';
+import { navigateTo, getNewPath, useQuery } from '@woocommerce/navigation';
+import { recordEvent } from '@woocommerce/tracks';
+import { Status } from '@wordpress/notices';
+
+/**
+ * Internal dependencies
+ */
+import ProductCard from '~/marketplace/components/product-card/product-card';
```
Comment (raicem) [inline plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx:15]: I'm going to visit all files and replace the usage of `~` with the relative imports.

File: plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx [hunk 3]
```diff
@@ -1,46 +1,92 @@
...
+			url: getNewPath( {
+				installProduct: product.id,
+			} ),
+		} );
+	}
+
+	// We hardcode this for now while we only display prices in USD.
+	const currencySymbol = '$';
+
+	if ( product.isInstallable && ! isInstalled ) {
```
Comment (raicem) [inline plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx:46]: I think we need one more check here: seeing this button doesn't guarantee that a user can install it. Installing a plugin depends on the `install_plugins` permission. While we check `manage_woocommerce` to load the page and authenticate APIs, there might a chance that the `install_plugins` cap may not be handed out to the store manager.

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx [hunk 4]
```diff
@@ -133,6 +134,7 @@ export default function Content(): JSX.Element {
 
 	return (
 		<div className="woocommerce-marketplace__content">
+			<InstallNewProductModal products={ products } />
```
Comment (raicem) [inline plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:137]: Currently, this flow doesn't work on the Featured/Discover page. Because there is no `Product[]` in the top level. I need to modify the data (it comes in groups and not a flat list), lift it up to `<Content>` and pass it down to the `InstallNewProductModal`. This is going to be a separate PR.

File: plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+			product_id: product.id,
+		} );
+
+		setInstallStatus( InstallFlowStatus.installing );
+
+		createOrder( product.id )
+			.then( ( response ) => {
+				// This narrows the CreateOrderResponse type to CreateOrderSuccessResponse
+				if ( ! response.success ) {
+					throw response;
```
Comment (raicem) [inline plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx:157]: Throwing this bubbles up the error and it caught by the catch block at the very end. That being said, I think code will never run, because when we're in the `then` block, `response.success` will always be true.

File: plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+			installStatus === InstallFlowStatus.installedCanActivate ||
+			installStatus === InstallFlowStatus.activating
+		) {
+			return __(
+				'Extension successfully installed. Would you like to activate it?',
+				'woocommerce'
+			);
+		}
+
+		if ( installStatus === InstallFlowStatus.installedCannotActivate ) {
```
Comment (raicem) [inline plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx:254]: Sometimes WordPress doesn't allow us to activate the plugin, by omitting the `activateUrl`. It can either:
- Plugin is already active: this happens when plugin is deleted without deactivation and then re-installed. 
- User doesn't have the right capability. 

I guide user to the plugins page in those cases. If you look at the `primaryButton` you'll see the main action is **View Plugins**

File: plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx (hunk 1 above)
Comment (bgrgicak) [reply plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx]: Yes, this will keep things consistent with the subscriptions page.
Any reason to not do it?

File: plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx (hunk 3 above)
Comment (bgrgicak) [reply plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx:46]: Could we add this check? https://github.com/woocommerce/woocommerce/blob/7068c78101a6062e66d46be0dedfc96ca86102fb/plugins/woocommerce-admin/client/settings-recommendations/recommendations-eligibility-wrapper.tsx#L26

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx (hunk 4 above)
Comment (bgrgicak) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:137]: I would leave this as optional in case we don't have time.

File: plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx (hunk 5 above)
Comment (bgrgicak) [reply plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx:158]: Why not use `Promise. reject()`?

File: plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx (hunk 1 above)
Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx]: No, I asked just to reduce the code as much as possible. I'm happy to leave it in.

File: plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx [hunk 7]
```diff
@@ -1,46 +1,92 @@
...
-	price: number | null | undefined;
-	currencySymbol: string;
-	averageRating: number | null | undefined;
-	reviewsCount: number | null | undefined;
+	product: Product;
 }
 
 function ProductCardFooter( props: ProductCardFooterProps ) {
-	const { price, currencySymbol, averageRating, reviewsCount } = props;
+	const { product } = props;
```
Comment (bgrgicak) [inline plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx:17]: Would be nice to move this into a component.

File: plugins/woocommerce-admin/client/marketplace/contexts/types.ts [hunk 8]
```diff
@@ -7,6 +7,7 @@ import { Options } from '@wordpress/notices';
  * Internal dependencies
  */
 import { Subscription } from '../components/my-subscriptions/types';
+import { Product } from '../components/product-list/types';
```
Comment (bgrgicak) [inline plugins/woocommerce-admin/client/marketplace/contexts/types.ts]:
```suggestion
```

File: plugins/woocommerce-admin/client/marketplace/utils/functions.tsx [hunk 9]
```diff
@@ -234,6 +236,16 @@ function disconnectProduct( subscription: Subscription ): Promise< void > {
 	} );
 }
 
+type WpAjaxReponse = {
```
Comment (bgrgicak) [inline plugins/woocommerce-admin/client/marketplace/utils/functions.tsx:239]: Should we make these types generally available in Woo? For example here `plugins/woocommerce-admin/client/typings/global.d.ts`?

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+			array(
+				'methods'             => 'POST',
+				'callback'            => array( __CLASS__, 'create_order' ),
+				'permission_callback' => array( __CLASS__, 'get_permission' ),
+				'args'                => array(
+					'product_id' => array(
+						'required'          => true,
+						'validate_callback' => function( $argument ) {
+							return is_int( $argument );
+						},
```
Comment (bgrgicak) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php:44]:
```suggestion
						'validate_callback' => 'is_int',
```

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php [hunk 11]
```diff
@@ -0,0 +1,72 @@
+<?php
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+class WC_Helper_Orders_API {
```
Comment (bgrgicak) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php:17]: Could we extend `WC_Helper_Subscriptions_API` here to reuse `get_permission`?

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+			);
+
+			return new \WP_REST_Response(
+				json_decode( wp_remote_retrieve_body( $response ), true ),
+				wp_remote_retrieve_response_code( $response )
+			);
+		} catch ( Exception $e ) {
+			return new \WP_REST_Response(
+				array(
+					'message' => 'Could not start the installation process. Reason: ' . $e->getMessage(),
```
Comment (bgrgicak) [inline plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php]:
```suggestion
					'message' => __(  'Could not start the installation process. Reason: ', 'woocommerce'  ) . $e->getMessage(),
```

**Comment by bgrgicak** [comment]
I left some comments and will do a full review soon. All the comments feel optional.

**Comment by bgrgicak** [comment]
> For back-end, make modifications in class-wc-helper. Make sure you also connect to the right Docker network.

@raicem could you please clarify this step?

I started testing on`add/19114-order-creation-helper-endpoint` (Woo.com), but wasn't being redirected, so I switched to trunk and it started working.

But now I get a 404 on `/wc/v3/marketplace/create-order`, I will return to testing this tomorrow morning.

File: plugins/woocommerce-admin/client/marketplace/utils/functions.tsx (hunk 9 above)
Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/utils/functions.tsx:239]: `WpAjaxResponseData` is not really complete. Also this calls `window.wp.updates.ajax` and I'm thinking response might vary a lot, depending on the action. Can we type that properly?

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php (hunk 11 above)
Comment (raicem) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php:17]: I directly called `WC_Helper_Subscriptions_API::get_permission()`. I'd like to avoid extending it.

File: plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php (hunk 10 above)
Comment (raicem) [reply plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php:44]: Oddly, this gives me `"Invalid parameter(s): product_id"`. No further info. I guess this doesn't pass the argument into the `is_int` function (just a guess).

File: plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx (hunk 3 above)
Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx:46]: Oh nice. I'll try this tomorrow and report back

File: plugins/woocommerce-admin/client/marketplace/components/content/content.tsx (hunk 4 above)
Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:137]: OK, thanks!

File: plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx (hunk 7 above)
Comment (raicem) [reply plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx:17]: Hmmm, not sure if I understand it. What do you want to move into a component?

**Comment by raicem** [comment]
> I started testing onadd/19114-order-creation-helper-endpoint (Woo.com), but wasn't being redirected, so I switched to trunk and it started working.

That's odd, I didn't merge the Woo.com branch, so there is no code to redirect. The redirection is made depending on the `create-order` endpoint response. I'm happy to help you tomorrow morning. 

> could you please clarify this step?
Sorry, I didn't give any details and it sounded cryptic. I meant tweaks we always make to the [$api_base ](https://github.com/woocommerce/woocommerce/blob/23db3eff69ad4606773dfd86e2ce0e06bf4dc874/plugins/woocommerce/includes/admin/helper/class-wc-helper-api.php#L33)and tweaking the request by hooks. I'll send you more details.

**Review by bgrgicak — APPROVED** [review]
This is amazing, thank you @raicem 🚀
